lib/spack/spack/spec.py

```diff
@@ -1815,38 +1813,32 @@ def node_dict_with_hashes(self, hash=ht.dag_hash):
         node = self.to_node_dict(hash)
         node[ht.dag_hash.name] = self.dag_hash()
 
-        # full_hash and build_hash are lazily computed -- but if we write
-        # a spec out, we want them to be included. This is effectively
-        # the last chance we get to compute them accurately.
+        # dag_hash is lazily computed -- but if we write a spec out, we want it
+        # to be included. This is effectively the last chance we get to compute
+        # it accurately.
         if self.concrete:
-            # build and full hashes can be written out if:
-            # 1. they're precomputed (i.e. we read them from somewhere
-            #    and they were already on the spec
-            # 2. we can still compute them lazily (i.e. we just made them and
+            # dag_hash can be written out if:
+            # 1. it's precomputed (i.e. we read it from somewhere
+            #    and it was already on the spec)
+            # 2. we can still compute it lazily (i.e. we just made the spec and
             #    have the full dependency graph on-hand)
             #
-            # we want to avoid recomputing either hash for specs we read
+            # we want to avoid recomputing the dag_hash for specs we read
             # in from the DB or elsewhere, as we may not have the info
             # (like patches, package versions, etc.) that we need to
-            # compute them. Unknown hashes are better than wrong hashes.
-            write_full_hash = (
-                self._hashes_final and self._full_hash or   # cached and final
-                not self._hashes_final)                     # lazily compute
-            if write_full_hash:
-                node[ht.full_hash.name] = self.full_hash()
-
-            write_build_hash = 'build' in hash.deptype and (
-                self._hashes_final and self._build_hash or  # cached and final
+            # compute it. Unknown hashes are better than wrong hashes.
+            write_dag_hash = (
+                self._hashes_final and self._hash or   # cached and final
                 not self._hashes_final)                     # lazily compute
-            if write_build_hash:
-                node[ht.build_hash.name] = self.build_hash()
+            if write_dag_hash:
+                node[ht.dag_hash.name] = self.dag_hash()
```

Comment:
I think this whole block can be reduced to:

```python
if self.concrete:
    node[ht.dag_hash.name] = self.dag_hash()
else:
    node['concrete'] = False
```

and we can get rid of all the `hashes_final` stuff now.